### PR TITLE
Readme update for link-bandwidth handling

### PR DESCRIPTION
**Pull request by rszarecki**

### CODE REVIEW

File: feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md [hunk 1]
```diff
@@ -19,43 +19,117 @@ Validate BGP in multipath UCMP support with link bandwidth community
...
+    *   Enable multipath, set maximum-paths limit to 2, enable allow multiple
+        AS, and send community type to [STANDARD, EXTENDED, LARGE]
+        *   /network-instances/network-instance/protocols/protocol/bgp/peer-groups/peer-group/afi-safis/afi-safi/use-multiple-paths/config/enabled
+        *   [TODO] /network-instances/network-instance/protocols/protocol/bgp/global/afi-safis/afi-safi/use-multiple-paths/ebgp/config/allow-multiple-as
+        *   [TODO] /network-instances/network-instance/protocols/protocol/bgp/globalp/afi-safis/afi-safi/use-multiple-paths/ebgp/config/maximum-paths
+        *   /network-instances/network-instance/protocols/protocol/bgp/peer-groups/peer-group/config/send-community-type
+        *   /network-instances/network-instance/protocols/protocol/bgp/global/use-multiple-paths/ebgp/link-bandwidth-ext-community/config/enabled
+    *   Advertise equal cost paths from port2 and port3 of ATE
+    *   Check entries in FIB for advertised prefix, it should have 2 entries.\
+        [TODO] The `weight` leafs of next-hops shall be in 2:1 ratio.
```
Comment (dplore) [inline feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md:44]: Can you provide an example of the expected data?

File: feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md [hunk 2]
```diff
@@ -19,43 +19,117 @@ Validate BGP in multipath UCMP support with link bandwidth community
...
+        *   /network-instances/network-instance/afts/next-hop-groups/next-hop-group/next-hops/state/index
+        *   /network-instances/network-instance/afts/next-hop-groups/next-hop-group/next-hops/state/weight
+    *   Initiate traffic from ATE port-1 to the DUT and destined to internal
+        target network
+    *   [TODO] Check entire traffic should only be unequally forwarded between DUT
+        port2 and port3 in ratio 2:1
```
Comment (dplore) [inline feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md:50]: Can you be more specific?   It's port 2 that should have 66% of traffic and port 3 is 33% of traffic (+/- 5%).  (or your choice of tolerance)

File: feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md [hunk 3]
```diff
@@ -19,43 +19,117 @@ Validate BGP in multipath UCMP support with link bandwidth community
...
+
+*   [TODO] RT-1.52.2: Verify use of equal community type
+
+    *   Configure ATE port 1, 2 and 3 on different AS, with boths AFI/SAFI
```
Comment (dplore) [inline feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md:54]: I suggest for the configuration generation portion of this step, say:

use config from RT-1.52.1, but change the link bandwidth communities as follows:
        * For ATE port 2 attach `link-bandwidth:23456:10K` extended-community
        * For ATE port 3 attach `link-bandwidth:23456:10K` extended-community

This simplifies the text and likely mirrors how the code would be written

File: feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md [hunk 4]
```diff
@@ -19,43 +19,117 @@ Validate BGP in multipath UCMP support with link bandwidth community
...
+        *   /network-instances/network-instance/afts/next-hop-groups/next-hop-group/next-hops/state/index
+        *   /network-instances/network-instance/afts/next-hop-groups/next-hop-group/next-hops/state/weight
+    *   Initiate traffic from ATE port-1 to the DUT and destined to internal
+        target network
+    *   Check entire traffic should only be unequally forwarded between DUT
+        port2 and port3 in ratio 1:1
+
+*   [TODO] RT-1.52.3: Verify BGP multipath when some path missing link-bandwidth extended-community
+
+    *   Configure ATE port 1, 2 and 3 on different AS, with boths AFI/SAFI
```
Comment (dplore) [inline feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md:77]: recommend same at RT-1.52.2 regarding config generation.  Just emphasize what has changed.

File: feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md [hunk 5]
```diff
@@ -19,43 +19,117 @@ Validate BGP in multipath UCMP support with link bandwidth community
...
+*   Create an single IPv4 internal target network attached to ATE port 2 and 3
+*   [TODO] Create an single IPv6 internal target network attached to ATE port 2 and 3
+
 
 ### Tests
 
-*   RT-1.52.1: Verify use of community type
+*   RT-1.52.1: Verify use of unequal community type
+
+    *   Configure ATE port 1, 2 and 3 on different AS, with boths AFI/SAFI
```
Comment (dplore) [inline feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md:31]: I recommend breaking this into two sections:
* Config generation + gnmi.Set REPLACE
* Validate (traffic generate and check counters) section.

Then you can more clearly refer to the config and 'reuse'  it in the following subtests

File: feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md [hunk 6]
```diff
@@ -19,43 +19,117 @@ Validate BGP in multipath UCMP support with link bandwidth community
...
+    *   Check entries in FIB for advertised prefix, it should have 2 entries.\
+        The `weight` leafs of next-hops shall be in 1:1 ratio.
+        *   /network-instances/network-instance/afts/next-hop-groups/next-hop-group/next-hops/state/index
+        *   /network-instances/network-instance/afts/next-hop-groups/next-hop-group/next-hops/state/weight
+    *   Initiate traffic from ATE port-1 to the DUT and destined to internal
+        target network
+    *   Check entire traffic should only be unequally forwarded between DUT
+        port2 and port3 in ratio 1:1
+
+*   [TODO] RT-1.52.4: Verify BGP multipath when none of paths has link-bandwidth extended-community
```
Comment (dplore) [inline feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md:98]: Is this a duplicate of RT-1.51?

https://github.com/openconfig/featureprofiles/tree/9bf506f64578e39a185c4f0cb7386c453e1c3942/feature/bgp/multipath/otg_tests/bgp_multipath_ecmp_test

File: feature/experimental/bgp/otg_tests/link_bandwidth_test/README.md [hunk 7]
```diff
@@ -133,6 +133,34 @@ bandwidth communities to routes based on a prefix match.
...
+      *
+    * Use gnmi Set REPLACE option for:
+      * `/routing-policy/policy-definitions` to configure the policy
+      * Use `/network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/export-policy`
+        to apply the policy on the DUT bgp neighbor to the ATE port 2.
+    * Verify expected communities are present in ATE port 2.
+      * Mark test as passing if Global Administartive valuee (ASN) of link-bakdwidth extended community **send by DUT** is either `23456` or ASN of DUT. 
+
+    * Expected community values for each policy
+      |              | set_linkbw_0                           | not_match_100_set_linkbw_1M |
```
Comment (dplore) [inline feature/experimental/bgp/otg_tests/link_bandwidth_test/README.md:148]: Add line break after the bullet item / before the header row to get table to render in markdown.

* bullet item

| Header | Header | Header |
|--------|--------|--------|
| Cell | Cell | Cell |
| Cell | Cell | Cell |

File: feature/experimental/bgp/otg_tests/link_bandwidth_test/README.md
Comment (dplore) [inline feature/experimental/bgp/otg_tests/link_bandwidth_test/README.md:1]: Can we remove the old deprecated requirements?  With revision control here in git, we can see the history to find deprecated items.

File: feature/experimental/bgp/otg_tests/link_bandwidth_test/README.md [hunk 8]
```diff
@@ -133,6 +133,34 @@ bandwidth communities to routes based on a prefix match.
...
+      | ------------ | -------------------------------------- | --------------------------- |
+      | prefix-set-1 | *DEPRECATED*                           | [none]                      |
+      | prefix-set-2 | *DEPRECATED*                           | [ "100:100" ]               |
+      | prefix-set-3 | *DEPRECATED*                           | [ "link-bandwidth:23456:0" ]  |
```
Comment (dplore) [inline feature/experimental/bgp/otg_tests/link_bandwidth_test/README.md:152]:
```suggestion
      | prefix-set-3 | *DEPRECATED*                           | [ "link-bandwidth:23456:1000000" ]  |
```

File: feature/experimental/bgp/otg_tests/link_bandwidth_test/README.md
Comment (dplore) [inline feature/experimental/bgp/otg_tests/link_bandwidth_test/README.md:1]: Please move this file out of the experimental folder so it does not conflict with this PR which is about to merge:

https://github.com/openconfig/featureprofiles/pull/3488

File: feature/experimental/bgp/otg_tests/link_bandwidth_test/README.md [hunk 9]
```diff
@@ -133,6 +133,34 @@ bandwidth communities to routes based on a prefix match.
...
+      * `/routing-policy/policy-definitions` to configure the policy
+      * Use `/network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/export-policy`
+        to apply the policy on the DUT bgp neighbor to the ATE port 2.
+    * Verify expected communities are present in ATE port 2.
+      * Mark test as passing if Global Administartive valuee (ASN) of link-bakdwidth extended community **send by DUT** is either `23456` or ASN of DUT. 
+
+    * Expected community values for each policy
+      |              | set_linkbw_0                           | not_match_100_set_linkbw_1M |
+      | ------------ | -------------------------------------- | --------------------------- |
+      | prefix-set-1 | *DEPRECATED*                           | [none]                      |
```
Comment (dplore) [inline feature/experimental/bgp/otg_tests/link_bandwidth_test/README.md:150]: For the not_match_100_set_linkbw_1M policy, if there is a route with no communities, should the policy result in adding the link-bandwidth community?  I think yes?

**Review by dplore — COMMENTED** [review]
Create a github issue tracking development for these TODO items.  Add a link to the github issue into this README.  (One issue is ok for all the TODO's)

File: feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md (hunk 6 above)
Comment (rszarecki) [reply feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md:98]: indeed. removed.

File: feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md (hunk 4 above)
Comment (rszarecki) [reply feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md:77]: done

File: feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md (hunk 3 above)
Comment (rszarecki) [reply feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md]: done

File: feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md [hunk 10]
```diff
@@ -28,95 +28,82 @@ Validate BGP in multipath UCMP support with link bandwidth community
...
+            *   Use UDP traffic with src and dst port randomly selected from 1-65535 range for each packet. Or equivalent pattern guaranteeng high entropy of traffic.
+    * Behaviour Validation
+        *   Check entries in AFT for advertised prefix, it should have 2 entries.\
+            [TODO] The `weight` leafs of next-hops shall be in 2:1 ratio.
+            *   Find next-hop-group IDs for both internal target networks:
+                *   /network-instances/network-instance/afts/ipv4-unicast/ipv4-entry[prefix=IPv4 internal target network]/state/**next-hop-group**
+                *   /network-instances/network-instance/afts/ipv4-unicast/ipv4-entry[prefix=IPv6 internal target network]/state/**next-hop-group**
+            *   using next-hop-group as key find number and weight of next-hops of both internal target network
+                *   /network-instances/network-instance/afts/next-hop-groups/next-hop-group[id=next-hop-group ID]/next-hops/state/index
+                *   /network-instances/network-instance/afts/next-hop-groups/next-hop-group[id=next-hop-group ID]/next-hops/state/**weight**
```
Comment (dplore) [inline feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md:55]:
```suggestion
                *   /network-instances/network-instance/afts/next-hop-groups/next-hop-group[id=next-hop-group ID]/next-hops/next-hop/state/index
                *   /network-instances/network-instance/afts/next-hop-groups/next-hop-group[id=next-hop-group ID]/next-hops/next-hop/state/**weight**
```

File: feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md [hunk 11]
```diff
@@ -28,95 +28,82 @@ Validate BGP in multipath UCMP support with link bandwidth community
...
+                * For ATE port 3 attach `link-bandwidth:23456:10K` extended-community
+    *   Behaviour Validation
+        *   Check entries in AFT for advertised prefix, it should have 2 entries.\
+            [TODO] The `weight` leafs of next-hops shall be in 1:1 ratio.
+            *   Find next-hop-group IDs for both internal target networks:
+                *   /network-instances/network-instance/afts/ipv4-unicast/ipv4-entry[prefix=IPv4 internal target network]/state/**next-hop-group**
+                *   /network-instances/network-instance/afts/ipv4-unicast/ipv4-entry[prefix=IPv6 internal target network]/state/**next-hop-group**
+            *   using next-hop-group as key find number and weight of next-hops of both internal target network
+                *   /network-instances/network-instance/afts/next-hop-groups/next-hop-group[id=next-hop-group ID]/next-hops/state/index
+                *   /network-instances/network-instance/afts/next-hop-groups/next-hop-group[id=next-hop-group ID]/next-hops/state/**weight**
```
Comment (dplore) [inline feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md:77]:
```suggestion
                *   /network-instances/network-instance/afts/next-hop-groups/next-hop-group[id=next-hop-group ID]/next-hops/next-hop/state/index
                *   /network-instances/network-instance/afts/next-hop-groups/next-hop-group[id=next-hop-group ID]/next-hops/next-hop/state/**weight**
```

File: feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md [hunk 12]
```diff
@@ -28,95 +28,82 @@ Validate BGP in multipath UCMP support with link bandwidth community
...
+            * For ATE port 3 **DO NOT** attach any link-bandwidth extended-community
+    *   Behaviour Validation
+        *   Check entries in AFT for advertised prefix, it should have 2 entries.\
+            [TODO] The `weight` leafs of next-hops shall be in 1:1 ratio.
+            *   Find next-hop-group IDs for both internal target networks:
+                *   /network-instances/network-instance/afts/ipv4-unicast/ipv4-entry[prefix=IPv4 internal target network]/state/**next-hop-group**
+                *   /network-instances/network-instance/afts/ipv4-unicast/ipv4-entry[prefix=IPv6 internal target network]/state/**next-hop-group**
+            *   using next-hop-group as key find number and weight of next-hops of both internal target network
+                *   /network-instances/network-instance/afts/next-hop-groups/next-hop-group[id=next-hop-group ID]/next-hops/state/index
+                *   /network-instances/network-instance/afts/next-hop-groups/next-hop-group[id=next-hop-group ID]/next-hops/state/**weight**
```
Comment (dplore) [inline feature/bgp/multipath/otg_tests/bgp_multipath_wecmp_lbw_community_test/README.md:100]:
```suggestion
            *   using next-hop-group as key find number and weight of next-hops of both internal target network
                *   /network-instances/network-instance/afts/next-hop-groups/next-hop-group[id=next-hop-group ID]/next-hops/next-hop/state/index
                *   /network-instances/network-instance/afts/next-hop-groups/next-hop-group[id=next-hop-group ID]/next-hops/next-hop/state/**weight**
```

**Review by dplore — COMMENTED** [review]
Minor comments on paths added.

ALso, please add a github issue for the TODO item, add it here in the README and also to the appropriate [github project](https://github.com/orgs/openconfig/projects)

**Review by dplore — APPROVED** [review]

**Pull request merged**